### PR TITLE
ci(e2e): add e2e test ci

### PR DESCRIPTION
**Pull request by anda-ren**
## Description
- add e2e test ci
## Modules
- [ ] UI
- [ ] Controller
- [ ] Agent
- [ ] Client
- [ ] Python-SDK
- [x] Others

## Checklist
- [ ] run code format and lint check
- [ ] add unit test
- [ ] add necessary doc

### CODE REVIEW

File: scripts/e2e_test/copy_artifacts_to_server.sh [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash -x
+
+source venv/bin/activate
+swcli instance login http://$1 --username starwhale --password abcd1234 --alias pre-k8s
```
Comment (tianweidut) [inline scripts/e2e_test/copy_artifacts_to_server.sh:4]: use other alias name?

**Comment by jialeicui** [comment]
`run_demo.sh` LGTM

File: scripts/e2e_test/copy_artifacts_to_server.sh (hunk 1 above)
Comment (anda-ren) [reply scripts/e2e_test/copy_artifacts_to_server.sh:4]: modified to `source $WORKDIR/source venv/bin/activate`

File: scripts/e2e_test/start_test.sh [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+
+in_github_action() {
+  [ -n "$GITHUB_ACTION" ]
+}
+
+file_exists() {
+  [ -f "$1" ]
+}
+
+  if in_github_action; then
```
Comment (tianweidut) [inline scripts/e2e_test/start_test.sh:11]:
``` bash
  if in_github_action; then
      export SW_PYPI_EXTRA_INDEX_URL='https://pypi.org/simple'
  else
      SW_PYPI_EXTRA_INDEX_URL='https://pypi.doubanio.com/simple/'
      export PARENT_CLEAN=true
  fi
```
Does the above code block has a right indention?

File: scripts/e2e_test/start_test.sh [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+username = $NEXUS_USER_NAME
+password = $NEXUS_USER_PWD
+EOF
+
+  cat $HOME/.pypirc
+}
+
+overwrite_pip_config() {
+  if file_exists "$HOME/.pip/pip.conf" ; then
+    cp $HOME/.pip/pip.conf $HOME/.pip/pip.conf.bake2etest
```
Comment (tianweidut) [inline scripts/e2e_test/start_test.sh:117]: `pip.conf.bake2etest ` --> `pip.conf.bak_e2e` ?

File: docker/external/config-e2e-pypi-repo.sh [hunk 4]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash -x
```
Comment (tianweidut) [inline docker/external/config-e2e-pypi-repo.sh:1]: `#!/bin/bash -x` -->

```bash
#!/usr/bin/env bash
set -e

if [[ ! -z ${DEBUG} ]]; then
    set -x
fi
```

- `set -e` : if the script meets some error, the script will exit.
- `set -x`: print verbose info

File: docker/Dockerfile.starwhale [hunk 5]
```diff
@@ -4,6 +4,13 @@ FROM ${BASE_IMAGE}
 ARG SW_VERSION=0.1.0
 ENV SW_VERSION ${SW_VERSION}
 
+ARG NEXUS_HOSTNAME=host.nexus
```
Comment (tianweidut) [inline docker/Dockerfile.starwhale:7]: in [sw-docker-entrypoint](https://github.com/star-whale/starwhale/blob/main/client/scripts/sw-docker-entrypoint#L28), there are some pypi config environment fields.  When running a container,  you can set `SW_PYPI_INDEX_URL`, `SW_PYPI_EXTRA_INDEX_URL` and `SW_PYPI_TRUSTED_HOST` envs.

Comment (tianweidut) [reply docker/Dockerfile.starwhale:7]: line 7 ~ line11 omitted?

File: docker/external/config-e2e-pypi-repo.sh (hunk 4 above)
Comment (tianweidut) [reply docker/external/config-e2e-pypi-repo.sh]: Other bash scripts have the same issue.

File: docker/Dockerfile.starwhale (hunk 5 above)
Comment (anda-ren) [reply docker/Dockerfile.starwhale:7]: when build docker image, this pip.conf is used too.

File: docker/external/config-e2e-pypi-repo.sh (hunk 4 above)
Comment (anda-ren) [reply docker/external/config-e2e-pypi-repo.sh]: fixed

File: .github/workflows/e2e-test.yml [hunk 6]
```diff
@@ -121,131 +62,11 @@ jobs:
...
-      - name: wait docker service up
-        uses: ifaxity/wait-on-action
-        with:
-          resource: http-get://localhost:8082
-
-      - name: run swcli stand alone e2e
-
-      - name: run server e2e
-
+          GITHUB_ACTION: 1
```
Comment (tianweidut) [inline .github/workflows/e2e-test.yml:70]: Remove `GITHUB_ACTION` env?   [Github action](https://docs.github.com/en/actions/learn-github-actions/environment-variables) will inject the env automatically.

Comment (anda-ren) [reply .github/workflows/e2e-test.yml]: done

File: docker/Dockerfile.starwhale (hunk 5 above)
Comment (tianweidut) [reply docker/Dockerfile.starwhale:7]: ok.

File: docker/external/config-e2e-pypi-repo.sh [hunk 7]
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env bash
+
+set -e
+
+if [[ ! -z ${DEBUG} ]]; then
+    set -x
+fi
+
+if [ $1 = false ] ; then
```
Comment (tianweidut) [inline docker/external/config-e2e-pypi-repo.sh:9]: `if [ $1 = false ] ; then` --> `if [ ! $1] ; then` ?

File: scripts/e2e_test/start_test.sh [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+
+start_nexus() {
+  docker run -d --publish=$PORT_NEXUS:$PORT_NEXUS --publish=$PORT_NEXUS_DOCKER:$PORT_NEXUS_DOCKER --name nexus  -e NEXUS_SECURITY_RANDOMPASSWORD=false $NEXUS_IMAGE
+  sudo cp /etc/hosts /etc/hosts.bak_e2e
+  sudo echo "127.0.0.1 $NEXUS_HOSTNAME" | sudo tee -a /etc/hosts
+}
+
+build_swcli() {
+  if in_github_action; then
+      pip install --upgrade pip
```
Comment (tianweidut) [inline scripts/e2e_test/start_test.sh:51]: `python3 -m pip install` ?

Comment (anda-ren) [reply scripts/e2e_test/start_test.sh]: done

**Review by tianweidut — APPROVED** [review]
LGTM

**Pull request merged**